src/data/sym2.lean

```diff
@@ -0,0 +1,201 @@
+/-
+Copyright (c) 2020 Kyle Miller All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Kyle Miller.
+-/
+
+import data.quot
+import tactic
+open function
+
+/-!
+# The symmetric square
+
+This file defines symmetric squares, which is `α × α` modulo swapping.
+This is also known as the type of unordered pairs.
+
+
+## Notation
+
+The symmetric square has a setoid instance, so `⟦(a, b)⟧` denotes a
+term of the symmetric square.
+
+## Tags
+
+symmetric square, unordered pairs
+-/
+
+namespace sym2
+universe u
+variables {α : Type u}
+
+/--
+This is the relation capturing the notion of pairs equivalent up to permutations.
+-/
+inductive rel (α : Type u) : (α × α) → (α × α) → Prop
+| refl (x y : α) : rel (x, y) (x, y)
+| swap (x y : α) : rel (x, y) (y, x)
+
+attribute [refl] rel.refl
+
+@[symm] lemma rel.symm {x y : α × α} : rel α x y → rel α y x :=
+by { tidy, cases a, exact a, apply rel.swap }
+
+@[trans] lemma rel.trans {x y z : α × α} : rel α x y → rel α y z → rel α x z :=
+by { tidy, cases_matching* rel _ _ _; apply rel.refl <|> apply rel.swap }
+
+instance rel.setoid (α : Type u) : setoid (α × α) :=
+by { use rel α, tidy, apply rel.trans a a_1 }
+
+/--
+This is the type for the symmetric square (i.e., the type of unordered pairs).
+-/
+@[reducible]
+def sym2 (α : Type u) [h : setoid (α × α)] := quotient h
+
+lemma eq_swap {a b : α} : ⟦(a, b)⟧ = ⟦(b, a)⟧ :=
+by { rw quotient.eq, apply rel.swap }
+
+lemma other_unique (a b c : α) : ⟦(a, b)⟧ = ⟦(a, c)⟧ ↔ b = c :=
+begin
+  split,
+  intro h, rw quotient.eq at h, cases h; refl,
+  intro h, apply congr_arg _ h,
+end
+
+/--
+A type `α` is naturally included in the diagonal of `α × α`, this function gives the image
+of this diagonal in `sym2 α`.
+-/
+def incl (x : α) : sym2 α := ⟦(x, x)⟧
+
+/--
+Checks whether a given element of `sym2 α` is in the image of `incl`.
+-/
+def is_diag (z : sym2 α) : Prop :=
+quotient.rec_on z (λ y, y.1 = y.2) (by { tidy, cases p; refl, cases p; refl })
+
+def image_incl_is_diag (z : sym2 α) (h : is_diag z) : ∃ x, z = incl x :=
+by tidy
+
+lemma diag_to_eq (x y : α) (h : is_diag ⟦(x, y)⟧) : x = y :=
```

Comment:
Lemmas that are about elements of `α × α` are a bit more general when formulated using `z : α × α` and then using `z` and `z.1` and `z.2` for its components.
Not every element in `α × α` is definitionally equal to a pair of two elements.